server/app/services/program/ActiveAndDraftPrograms.java

```diff
@@ -45,31 +46,36 @@ public static ActiveAndDraftPrograms buildFromCurrentVersionsUnsynced(
     return new ActiveAndDraftPrograms(repository, Optional.empty());
   }
 
+  private ImmutableMap<String, ProgramDefinition> mapNameToProgramWithFilter(
+      VersionRepository repository,
+      Optional<ProgramService> service,
+      VersionModel versionModel,
+      Optional<DisplayMode> excludeDisplayMode) {
+    return repository.getProgramsForVersion(checkNotNull(versionModel)).stream()
+        .map(
+            program ->
+                service.isPresent()
+                    ? getFullProgramDefinition(service.get(), program.id)
+                    : program.getProgramDefinition())
+        .filter(
+            program ->
+                excludeDisplayMode.isPresent()
+                    ? program.displayMode() != excludeDisplayMode.get()
+                    : true)
+        .collect(ImmutableMap.toImmutableMap(ProgramDefinition::adminName, Function.identity()));
+  }
+
   private ActiveAndDraftPrograms(VersionRepository repository, Optional<ProgramService> service) {
     VersionModel active = repository.getActiveVersion();
     VersionModel draft = repository.getDraftVersionOrCreate();
     // Note: Building this lookup has N+1 query behavior since a call to getProgramDefinition does
     // an additional database lookup in order to sync the set of questions associated with the
     // program.
     ImmutableMap<String, ProgramDefinition> activeNameToProgram =
-        repository.getProgramsForVersion(checkNotNull(active)).stream()
-            .map(
-                program ->
-                    service.isPresent()
-                        ? getFullProgramDefinition(service.get(), program.id)
-                        : program.getProgramDefinition())
-            .collect(
-                ImmutableMap.toImmutableMap(ProgramDefinition::adminName, Function.identity()));
+        mapNameToProgramWithFilter(repository, service, active, Optional.empty());
```

Comment:
**suggestion:** You can create another function `mapNameToProgram` that calls `mapNameToProgramWithFilter(repository, service, active, Optional.empty());`. May make this look a little cleaner since you have an option in the cases where you don't want any filter applied. For use here and below.